Add meta function for channel_shuffle operation

Episkey0109 opened the pull request:
This commit introduces a meta function for the `channel_shuffle` operation, enabling PyTorch to perform shape inference and optimizations related to this operation without actual computation. The meta function assumes input shape (*, C, H, W).

Fixes #122771

Episkey0109 committed:
Add meta function for channel_shuffle operation

This commit introduces a meta function for the `channel_shuffle` operation, enabling PyTorch to perform shape inference and optimizations related to this operation without actual computation. The meta function assumes input shape (N, C, H, W) and validates that the number of channels (C) is divisible by the specified number of groups.

vadimkantorov commented:
The online docs assume any-dimension `(*, C, H, W)`, right?

Also, the code example in online docs seems incorrect as `torch.randn` should return float values around 0 (including negatives) and not integers like `15`:
![image](https://github.com/pytorch/pytorch/assets/1041752/14b5e846-6798-46f6-85a1-fefacf637591)

Episkey0109 commented:
Hi @vadimkantorov, thanks for pointing this out! I'm looking into fixing the dimension problem

vadimkantorov commented:
Also maybe a TeX issue in the docs about rendering C g underscore

Episkey0109 committed:
Refactor meta_channel_shuffle function to support input tensors with any number of leading dimensions

This commit updates the meta_channel_shuffle function to handle input tensors with arbitrary leading dimensions. Instead of assuming a fixed input shape of (N, C, H, W), the function now uses *leading_dims to capture any number of leading dimensions that may precede the last three dimensions (C, H, W).

Episkey0109 commented:
I think your suggestion makes a good point. Could you open a new issue so that someone else can look into this problem? Thanks!

Episkey0109 committed:
fix lint issue

mikaylagawarecki reviewed (commented):
Could you add an opinfo for channel_shuffle to [this file](https://github.com/pytorch/pytorch/blob/main/torch/testing/_internal/common_methods_invocations.py) so that the TestMeta tests decorated with `@ops` in [this file ](https://github.com/pytorch/pytorch/blob/main/test/test_meta.py) will run please

Episkey0109 committed:
Add OpInfo and sample inputs for nn.functional.channel_shuffle

This commit introduces an OpInfo entry for the `channel_shuffle` function in `nn.functional`, along with a `sample_inputs_channel_shuffle` function to generate valid inputs for systematic testing.

ezyang commented:
@pytorchbot merge

Episkey0109 commented:
@pytorchbot label "topic: not user facing"

Episkey0109 committed:
fix channel_shuffle OpInfo

ezyang commented:
@pytorchbot merge

mikaylagawarecki commented:
@Episkey0109 the failures look real, you might need to add a few skips for CI to be happy

Episkey0109 commented:
@mikaylagawarecki I have added skips to solve 3 failures on macos-12-py3-arm64 and win-vs2019-cpu-py3, but I’m confused by the 2 failures on linux-focal-cuda12.1-py3.10-gcc9:
```
profiler/test_profiler.py::TestProfiler::test_source_multithreaded_multiple_preexisting_work_in_main_thread_True - AssertionError: Scalars are not equal!
Expected 3 but got 4.
```
Is there anything I’m supposed to do to deal with this? I could not reproduce the failure locally.

mikaylagawarecki commented:
The profiler failures are unrelated, should not be anything on your end there

Episkey0109 committed:
Add skip

mikaylagawarecki commented:
@pytorchbot merge

Episkey0109 commented:
@pytorchbot merge

Episkey0109 commented:
```
Merge of https://github.com/pytorch/pytorch/pull/123033 failed due to: 1 mandatory check(s) are pending/not yet run.  The first few are:
- pull
```
@mikaylagawarecki seems like the pull workflow must be run and is still awaiting approval. Could you approve it?

mikaylagawarecki commented:
@pytorchbot merge -r

Episkey0109 committed:
fix OpInfo

Episkey0109 committed:
update meta func & add skips

ezyang commented:
@pytorchbot merge

Episkey0109 commented:
@pytorchbot merge
